allennlp/nn/beam_search.py

```diff
@@ -196,13 +195,23 @@ def new_step(
         for key, state_tensor in state.items():
             if state_tensor is None:
                 continue
-            _, *last_dims = state_tensor.size()
-            # shape: (batch_size * beam_size, *)
-            state[key] = (
-                state_tensor.unsqueeze(1)
-                .expand(batch_size, self.beam_size, *last_dims)
-                .reshape(batch_size * self.beam_size, *last_dims)
-            )
+
+            if state_tensor.dim() == 3 and key in ["decoder_hidden", "decoder_context"]:
```

Comment:
I use key name check here. 
As I mentioned: the `ComposedSeq2Seq` has `state['encoder_output'] with shape (batch, length, hidden)`.
It's hard to check whether it has the dimension of `num_of_layers`.
So the directly name check works. @dirkgr

Reply:
@Tobias-Rohde, as the last person that has been deep inside this code, do you have an opinion on this?

Reply:
I am not sure if I like adding this decoder specific code to the beam search.

Couldn't you instead change the model to transpose the state tensors to have the num_layers dimension at the end? So for instance for the SimpleSeq2Seq model, here:

https://github.com/allenai/allennlp-models/blob/4d0e090514af273471555be08676bc02999fd0ff/allennlp_models/generation/models/simple_seq2seq.py#L415

you would get the state tensors and transpose them such that num_layers is the first dimension and then before putting them back into the state here:

https://github.com/allenai/allennlp-models/blob/4d0e090514af273471555be08676bc02999fd0ff/allennlp_models/generation/models/simple_seq2seq.py#L438

you would transpose them to have num_layers as the last dimension. That way we could leave the beam search unmodified and move the decoder specific to the model, where it belongs imo. Let me know if this would work, otherwise I think the way you implemented it is fine too.

Reply:
Thanks for your quickly response. 
There are two different ways to work. I have discussed with @dirkgr  in [here](https://github.com/allenai/allennlp-models/pull/90#issuecomment-659231259).
I prefer the second way( make changes in beam search rather than model decoders).
The reason is  as follows: It is the beam_search code that requires state shape like `(batch,...)`. If you implement it in decoder model, that means you should repeat same code in every (LSTM)decoder.

Reply:
I see your point, I think both ways are fine. I'd say just keep it the way you've implemented it currently. The beam search code should be rewritten in the near future anyways, there are many very useful beam search features that huggingface and fairseq have (min length, max length, max repeated n-grams, ...) and I think we should have those too.

Reply:
Ok, hope for your future version~ 
@dirkgr  Can we merge it into master, then I can update my PR in `allennlp-model`